### PR TITLE
ggPMX Update

### PR DESCRIPTION
**Pull request by sehamsick**
Adding NONMEM reader and minor Bugfixes

### CODE REVIEW

**Comment by mattfidler** [comment]
Hi @sehamsick 

The checks do not pass on all systems.  We can discuss how to fix them.

File: R/read_nmext.R [hunk 1]
```diff
@@ -1,17 +1,28 @@
-#' Title
+# This code is strongly based on the read_nmext function of mrgsolve (nmxml.R) (Thanks to Kyle T Baron)
+# The code was slightly adjusted to check for multiple tables and also extract SE (ITERATION == 1000000001)
+# The output was also slightly adjust to fit ggPMX output (df and df2)
+# Addition of param, "quiet", (option of msg function, from xpose package)
+
+#' Reads .ext files generated by NONMEM
```
Comment (mattfidler) [inline R/read_nmext.R:6]: I would add a @author section here and give him credit, and you credit as well.  Also note where the differences are in this function.  One last point is that I would name it something different to avoid a name clash with mrgsolve (in case they use both).

File: DESCRIPTION [hunk 2]
```diff
@@ -17,16 +17,18 @@ Authors@R: c(
               person("Inga", "Ludwig", email="inga.ludwig@novartis.com", role="aut"),
               person("Ines", "Paule", email="ines.paule@novartis.com", role="aut"),
               person("Didier", "Renard", email="didier.renard@novartis.com", role="aut"),
-	      person("Matthew","Fidler", role = c("ctb", "cre"), email = "matthew.fidler@gmail.com", comment=c(ORCID="0000-0001-8538-6691")),
-	      person("Seid", "Hamzic", email = "se.hamzic@gmail.com",role ="aut"),
+	            person("Matthew","Fidler", role = c("ctb", "cre"), email = "matthew.fidler@gmail.com", comment=c(ORCID="0000-0001-8538-6691")),
+	            person("Seid", "Hamzic", email = "se.hamzic@gmail.com",role ="aut"),
               person("Benjamin", "Guiastrennec", email = "guiastrennec@gmail.com",role ="ctb"),
+              person("Kyle T", "Baron", email = "kyleb@metrumrg.com",role ="ctb"),
```
Comment (mattfidler) [inline DESCRIPTION:23]: Add his OCRID too:

https://github.com/metrumresearchgroup/mrgsolve/blob/f1c4b3a286d0bbe405c777c81b3fcd722dbb894e/DESCRIPTION#L10

File: R/read_nmext.R [hunk 3]
```diff
@@ -28,19 +39,24 @@ read_nmext <- function(run=NA_real_, project = getwd(), file=paste0(run,".ext"),
   
   use_dt <- requireNamespace("data.table",quietly=TRUE) & read_fun=="data.table"
   
-  ##Check for multiple tables
-  ext_tmp <- readr::read_lines(extfile,n_max=-1)
+  ## Check for multiple tables/problems
+  
+  ext_tmp <- readr::read_lines(extfile,n_max=-1) ## May be substituted for another read function, so package does not depend on readr
```
Comment (mattfidler) [inline R/read_nmext.R:44]: Use `fread` here please;  That way you don't add another dependency.

File: R/read_nmext.R [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      " while reading 'ext' file ", shQuote(basename(extfile))
+    )
+  }
+  
+  ans <- as.list(ans)
+  names(ans) <- gsub("[[:punct:]]", "", names(ans))
+  ans <- list(
+    param = ans[grepl("THETA", names(ans))],
+    omega = mrgsolve::as_bmat(ans, "OMEGA"), 
+    sigma = mrgsolve::as_bmat(ans, "SIGMA"),
```
Comment (mattfidler) [inline R/read_nmext.R:101]: If you already use `mrgsolve` functions within here without importing `mrgsolve`, perhaps you should copy them here and drop the dependency.  That way you wouldn't depend on a simulation package to read `NONMEM` files.  If you are in the monolix camp you may only want to use `simulx`.  For nonmem, maybe `PsN` or `NONMEM` directly. For `nlmixr` maybe `RxODE`.

The other advantage is if `mrgsolve` changes this code, your code wont break.

File: inst/examples/pmx_nm.R [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  directory=model_dir,
+  file = "run001.lst", 
+  simfile = "simulation.ctl"
+)
+
+## loading with individual table(s)-names
+ctr <- pmx_nm(directory = model_dir,
+              runno = 3, 
+              table_names = "xptab")
+}
```
Comment (mattfidler) [inline inst/examples/pmx_nm.R:26]: Why is this file here?  Usually examples are in vignettes or in documentation.  This is entirely outside and still contains some `Rd` markdown.  Is this a mistake?

File: tests/testthat/test-nonmem_reader.R [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  
+  #Check if simulation data was extracted correctly
+  simput_file <- file.path(csv_dir, "sim.csv")
+  sim_dat <- read.csv(simput_file)
+  
+  expect_equal(
+    nrow(ctr %>% get_data("sim")), nrow(sim_dat)
+    )
+  
+})
```
Comment (mattfidler) [inline tests/testthat/test-nonmem_reader.R:47]: Some platforms (like linux) can sometimes fail if you don't have a newline on the end of your code;  Can you fix this test?

File: vignettes/ggPMX-guide.Rmd [hunk 7]
```diff
@@ -337,7 +399,7 @@ ctr %>% plots()
 
 ### Datasets
 
-`ggPMX` uses the following dataset name convention. The *input modeling dataset* is the one used for model fitting (the actual data). The *output modeling datasets* are those output from the fitting tool (Monolix or NONMEM). The *ggPMX datasets* are the ones created within ggPMX. Table \ref{tab:ggPMX_datasets} provides a list of all ggPMX datasets.
+`ggPMX` uses the following dataset name convention. The *input modeling dataset* is the one used for model fitting (the actual data) (only for Monolix). The *output modeling datasets* are those output from the fitting tool (Monolix,NONMEM or nlmixr). The *ggPMX datasets* are the ones created within ggPMX. Table \ref{tab:ggPMX_datasets} provides a list of all ggPMX datasets.
```
Comment (mattfidler) [inline vignettes/ggPMX-guide.Rmd:402]: Is this comment about monolix for the (input modeling dataset) necessary?  I don't think it adds anything here.

File: vignettes/ggPMX-guide.Rmd [hunk 8]
```diff
@@ -467,6 +532,20 @@ ctr <- pmx_mlx(
...
+
+```{r, eval = FALSE}
+ctr <- pmx_nm(
+  directory = model_dir,
+  file      = "modelfile.ctl" #or .lst
+  simfile   = "simulation_modelfile.ctl" #or .lst
+)
+```
+
+
```
Comment (mattfidler) [inline vignettes/ggPMX-guide.Rmd:548]: Probably need a nlmixr section here;  It should be easy enough to add.

File: vignettes/ggPMX-guide.log [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+Here is how much of TeX's memory you used:
+ 12276 strings out of 495063
+ 178416 string characters out of 3182201
+ 287912 words of memory out of 3000000
+ 15385 multiletter control sequences out of 15000+200000
+ 4403 words of font info for 15 fonts, out of 3000000 for 9000
+ 14 hyphenation exceptions out of 8191
+ 36i,0n,38p,293b,272s stack positions out of 5000i,500n,10000p,200000b,50000s
+
+!  ==> Fatal error occurred, no output PDF file produced!
```
Comment (mattfidler) [inline vignettes/ggPMX-guide.log:597]: This should be removed from your vignettes directory;  Also it seems your PDF wasn't produced.

**Review by mattfidler — CHANGES_REQUESTED** [review]

File: tests/testthat/test-nonmem_reader.R (hunk 6 above)
Comment (sehamsick) [reply tests/testthat/test-nonmem_reader.R:47]: New empty line added (48)

File: R/read_nmext.R (hunk 1 above)
Comment (sehamsick) [reply R/read_nmext.R:6]: Added @author (also in pmx_nm), renamed function to "read_extfiles", added code line numbers to indicate where the changes are

File: DESCRIPTION (hunk 2 above)
Comment (sehamsick) [reply DESCRIPTION:23]: fixed

File: tests/testthat/test-nonmem_reader.R (hunk 6 above)
Comment (sehamsick) [reply tests/testthat/test-nonmem_reader.R:47]: empty line 48 added

File: vignettes/ggPMX-guide.log (hunk 9 above)
Comment (sehamsick) [reply vignettes/ggPMX-guide.log:597]: Yes, I don't know why it fails to knit the PDF... guide.log removed

File: inst/examples/pmx_nm.R (hunk 5 above)
Comment (sehamsick) [reply inst/examples/pmx_nm.R:26]: Other examples of documentation of other functions are also in this folder. I didn't want to change to much in the structure, we might discuss this

File: vignettes/ggPMX-guide.Rmd (hunk 8 above)
Comment (sehamsick) [reply vignettes/ggPMX-guide.Rmd:548]: Yes! It was on the todo list, but forgot to add it, now its in there!

File: vignettes/ggPMX-guide.Rmd (hunk 7 above)
Comment (sehamsick) [reply vignettes/ggPMX-guide.Rmd:402]: It's not necessary here, because it is explained further above. But in generall i would leave it, because it's used throughout the text. Might be changed in a future update, where we could revise more of the text.

For now I added NONMEM and nlmixr, and removed some of the errors

File: R/read_nmext.R (hunk 4 above)
Comment (sehamsick) [reply R/read_nmext.R:101]: I copied the code, however do i have to rename them as well? I didn't change anything to it

File: R/read_nmext.R (hunk 3 above)
Comment (sehamsick) [reply R/read_nmext.R:44]: I couldn't figure out how to use `fread`, however no i am using `readLines`, which is a base R function
